numba/np/arrayobj.py

```diff
@@ -981,7 +995,11 @@ def prepare(self):
         for i in self.indexers:
             i.prepare()
         # Compute the resulting shape
-        self.indexers_shape = sum([i.get_shape() for i in self.indexers], ())
+        shape_list = [i.get_shape() for i in self.indexers]
+        one = self.context.get_constant(types.intp, 1)
+        for i in self.newaxes:
+            shape_list.insert(i, (one,))
```

Comment:
This seems to be inserting at indexes based on a previous loop index into a list which has changing length. I guess this could work but it's quite hard to follow. This made me wonder if there's merit in implementing a `NewAxisIndexer` class as a way to handle the `newaxis` without there being a custom path and custom tracking/fixing of the shapes etc?

Reply:
I tried out this approach. The problem with it is, the way `Indexer`s built, they are supposed to provide an index for every axis in the 'source' array and not the resulting array. The number of `Indexer`s are supposed to correspond to the number of axes in the 'source'. (This is also represented by the following, which asserts that number of `Indexer` objects should always match the original dimensions of array.)

https://github.com/numba/numba/blob/bfe1270e23f8daf7fd95cfb6ef385f6c832fd927/numba/np/arrayobj.py#L1005

By adding a `np.newaxis`/`None`, we are adding a dimension in the resulting array and not the original one from where we're accessing the elements from. The original remains unchanged. 

So adding a `NewAxisIndexer`will leads to breaking of this `assert`s and requires special case handling in the index provision logic where we'll need to exclude the `NewAxisIndexer` objects since they don't have a existence in original array.

https://github.com/numba/numba/blob/bfe1270e23f8daf7fd95cfb6ef385f6c832fd927/numba/np/arrayobj.py#L1059

So it's possible, but will only add more complexity. Also we would want to take into consideration that currently the logic for both basic and advanced indexing is nearly the same. The above `Indexer`s will only apply to advanced indexing and not basic one. I think instead, we can somehow work on simplifying the current logic to make it less bug-prone/readable.

Reply:
Ah ok, thanks for considering it. I would expect `assert`s to need updating if logic changes, so I am not convinced that is an issue in itself, however additional/increased complexity is a problem and something to avoid. I think perhaps making the current logic have a more clear intent seems like the best thing to do?